_posts/2018-02-05-chaining-http-requests-in-elm.md

```diff
@@ -0,0 +1,342 @@
+    const user = await responseB.json()
+    if (!user.name) {
+        console.log("unspecified")
+    } else {
+        console.log(user.name)
+    }
+} catch (err) {
+    console.error(err)
+}
+```
+
+This is already complicated code to read and understand, and it's tricky to do using Elm as well. Let's see how to achieve the same understanding exactly what we're doing (we've all blindly copied and pasted code in the past, don't deny).
+
+First, let's write the two requests we need; one for fetching the list of events, the second to obtain a given user's details from her `login`:
+
+```haskell
+import Http
+import Json.Decode as Decode
+
+eventsRequest : Http.Request (List String)
+eventsRequest =
+    Http.get "https://api.github.com/events"
+        (Decode.at [ "actor", "login" ] Decode.string
+            |> Decode.list
+        )
+
+nameRequest : String -> Http.Request String
+nameRequest login =
+    Http.get ("https://api.github.com/users/" ++ login)
+        (Decode.at [ "name" ]
+            (Decode.oneOf
+                [ Decode.string
+                , Decode.null "unspecified"
+                ]
+            )
+        )
+```
+
+These two functions return `Http.Request` with the type of data they'll retrieve from the JSON body of their respective responses. `nameRequest` handles the case where Github users don't have entered their full name yet, so the `name` field might be a `null`; as with the JavaScript version, we then default to `"unspecified"`.
+
+That's good but now we need to execute and chain these two requests, the second one depending on the result of the first one, where we retrieve the `actor.login` value of the event object.
+
+Elm is a pure language, meaning you can't have side effects in your functions (an HTTP request is a huge side effect). So your functions must return *something* that represents a given side effect, then another function must handle the result when a response is received.
+
+In Elm, you're gonna use a [Task] to define and perform such asynchronous side effects, and tasks may succeed or fail (like Promises do in JavaScript). The [Http] package provides `Http.toTask` to map a request to a `Task`. Let's use that here:
+
+```haskell
+fetchEvents : Task Http.Error (List String)
+fetchEvents =
+    eventsRequest |> Http.toTask
+
+fetchName : String -> Task Http.Error String
+fetchName login =
+    nameRequest login |> Http.toTask
+```
+
+I created these two simple functions mostly to focus on their return types; a `Task` must define an error type and a result type. For example, `fetchEvents` being an HTTP task, it will receive an `Http.Error` when the task fails, and a list of strings when the task succeeds.
+
+But dealing granularily with HTTP errors being out of scope of this already long blog post, and in order to keep things as less convoluted as possible, I'm gonna use `Task.mapError` to turn complex HTTP error objects into strings as well:  
```

Comment:
Are you really sure this is out of scope? I don't like articles that are "almost" real life examples, but for the sake of simplicity do stuff that you wouldn't do in real life. Maybe showing how to deal with HTTP errors (just logging the status code in the console for example) may help people understand how to deal with `Tasks`?
This is just a suggestion, but to me, adding a "shortcut" for the sake of simplicity is often a deal breaker for beginners.

Reply:
Having a string representation of an HTTP error is not avoiding the topic at all, as it makes them immediately usable and readable in a single line of code. Handling HTTP errors is another article as long as this one.